### PR TITLE
refactoring!(share): migrating to go-square v2 types

### LINKED ISSUE #3770: [Feature Request]: Migrate to go-square types
### Implementation ideas

We have a new v2 go square [package](https://github.com/celestiaorg/go-square/releases/tag/v2.0.0) which is intended to unify the types across both the `celestia-node` and `celestia-app` package. As part of preparing to upgrade to v3, we need to migrate the existing core types like: 

- `Share`
- `Namespace`
- `Blob`
- `BlobTx`
- `Square`

to the new types specified in this repo.

I'm unfamiliar of how this may break the APIs for celestia-node.

V2 of go-square is backwards compatible from the square construction perspective. This means that all squares constructed using the v1 and v2 of the application can be constructed with the v2 go-square. The construction is **not** forwards compatible. v3 of celestia-app can produce a square (using v2 of go-square) which is different to v1 of go-square.

### PR DESCRIPTION
**Pull request by vgonkivs**
Fixes #3770 
Fixes https://github.com/celestiaorg/celestia-node/issues/3824

PLEASE NOTE: **We are not backward compatible since the app version has changed**

### CODE REVIEW

File: square/namespace.go
Comment (cristaloleg) [inline square/namespace.go:81]: Extra comment :)

File: go.mod
Comment (cristaloleg) [inline go.mod:349]: We already have v2.0.0 ;) https://github.com/celestiaorg/go-square/releases/tag/v2.0.0

File: blob/service.go
Comment (cristaloleg) [inline blob/service.go:392]: dead `if`

File: square/shwap/row.go
Comment (cristaloleg) [inline square/shwap/row.go:98]: Why this get dropped?

File: cmd/util_test.go
Comment (cristaloleg) [inline cmd/util_test.go:3]: Will be uncommented, right?

File: go.mod
Comment (vgonkivs) [reply go.mod:349]: I know. But it does not have my local changed ;)

File: square/shwap/row.go
Comment (vgonkivs) [reply square/shwap/row.go:98]: https://github.com/celestiaorg/go-square/blob/a446cee3ad5fd5be5e59dc9a027bc35b4f6100cb/share/share.go#L23

File: blob/blob.go
Comment (cristaloleg) [inline blob/blob.go:93]: Rename in this PR ?

File: square/shwap/row_namespace_data.go
Comment (cristaloleg) [inline square/shwap/row_namespace_data.go:183]: This PR ?

Comment (vgonkivs) [reply square/shwap/row_namespace_data.go:183]: it is already reworked. Forgot to remove the comment

File: square/namespace.go
Comment (vgonkivs) [reply square/namespace.go:80]: the file will be removed

File: share/shwap/sample.go
Comment (rootulp) [inline share/shwap/sample.go:126]:
```suggestion
// parity shares are located outside of first quadrant of the square. According to the nmt
```

File: store/file/ods.go
Comment (rootulp) [inline store/file/ods.go:99]:
```suggestion
// padding share. Tail padding share are constant and aren't stored.
```

File: store/file/ods.go
Comment (rootulp) [inline store/file/ods.go:231]:
```suggestion
	// axis is a column and from the left half of the square.
```

File: store/file/q4.go
Comment (rootulp) [inline store/file/q4.go:15]:
```suggestion
// q4 stores the fourth quadrant of the square.
```

File: store/file/square.go
Comment (rootulp) [inline store/file/square.go:17]:
```suggestion
// readSquare reads Shares from the reader and returns a square. It assumes that the reader is
```

**Review by rootulp — COMMENTED** [review]
Overall LGTM.

1. A lot of comments need to be reverted `share` => `square`
2. go-ci complains that there are Go compilation errors in some of the tests. E.g. nodebuilder/tests/da_test.go
3. I think we need to be careful about the `appVersion` constant because celestia-app will transition from app version 2 => 3. I'm pretty confident we can future-proof by making that constant 3 now because all versioned constants that existed in v2 have the same values in v3. However, v3 introduced new versioned constants.

File: share/availability/full/testing.go
Comment (vgonkivs) [reply share/availability/full/testing.go:27]: this code will be uncommented after all necessary changes are done

File: share/availability/light/testing.go
Comment (vgonkivs) [reply share/availability/light/testing.go:22]: Will be uncomented

File: blob/blob.go
Comment (vgonkivs) [reply blob/blob.go]: changed to v3.Version

**Review by walldiss — COMMENTED** [review]
The introduction of new types to RPC level introduces huge API break surface. This might be very painful for users to suddenly notice that most API endpoints types has changed. Alternative could be to keep old types in RPC, while change only internal node usage.

File: blob/blob.go
Comment (rootulp) [reply blob/blob.go]: Thanks!

**Comment by vgonkivs** [comment]
> The introduction of new types to RPC level introduces huge API break surface. This might be very painful for users to suddenly notice that most API endpoints types has changed. Alternative could be to keep old types in RPC, while change only internal node usage.

That's true, but on the other side I don't want to keep multiple types that declare the same thing in different packages. It might be painfull for the users to migrate to the new types but overall it will help to avoid inconsistency of the types. 
Please also note that the underlying types have also changed(from `[]byte` to `struct{[]byte}`)

File: share/root.go
Comment (renaynay) [inline share/root.go:94]: We should add an error to this function for the caller to understand when there's an error bc the caller could interpret this as a success case since 

if !IsOutsideRange is the call, then invalid length hashes could falsely pass that condition.

Comment (renaynay) [reply share/root.go:94]: This goes the same for the other functions here.

File: share/shwap/p2p/bitswap/getter.go
Comment (renaynay) [inline share/shwap/p2p/bitswap/getter.go:26]: doc

File: share/shwap/p2p/shrex/shrex_getter/shrex.go
Comment (renaynay) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go:94]: doc

File: share/shwap/p2p/shrex/shrexnd/client.go
Comment (renaynay) [inline share/shwap/p2p/shrex/shrexnd/client.go:49]: doc

File: share/shwap/p2p/shrex/shrexnd/doc.go
Comment (renaynay) [inline share/shwap/p2p/shrex/shrexnd/doc.go:34]: doc

File: share/shwap/p2p/shrex/shrexsub/pb/notification.proto
Comment (renaynay) [inline share/shwap/p2p/shrex/shrexsub/pb/notification.proto:3]: what?

**Review by renaynay — CHANGES_REQUESTED** [review]
Thank you for this PR it was a very heavy lift. There are a few issues I flagged in review comments, but mainly need shrexsub proto package name change to be reverted, and also need to make sure ext header proto change is backwards compat Please.

File: share/shwap/p2p/shrex/shrexsub/pb/notification.proto
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrexsub/pb/notification.proto:3]: accident

File: blob/blob.go
Comment (vgonkivs) [reply blob/blob.go:79]: a v1 blob contains also a signer. So, having NewBlob constructs only the latest blob types is not the case because new version can have new additional fields

Comment (vgonkivs) [reply blob/blob.go:79]: but we can have NewBlob + applied options

File: square/availability.go
Comment (vgonkivs) [reply square/availability.go]: already returned it back to share

File: blob/blob.go
Comment (Wondertan) [reply blob/blob.go:79]: The main reason to introduce a versioned constructor is devex. You don't want to import a constant from the app every time you need to make a share, a blob, or a namespace. Also, each version may have different set of params, like V0 Blob doesn't have a signer and V1 does, so you cant' necessary generalize the constructor. 

Furthermore, I would argue that users should be aware of its version. Depending always on the last version may lead to accidental breakages for the blob-consuming code after the type version bump.  I think its useful might be to provide `NewBlob` option that always points to latest version, but only along other versioned constructors.

File: blob/blobtest/testing.go
Comment (vgonkivs) [reply blob/blobtest/testing.go:39]: as it is used in `TestSkipPaddingsAndRetrieveBlob`

File: blob/parser.go
Comment (vgonkivs) [reply blob/parser.go:83]: Yes, this is true. I'll modify the condition.

File: blob/service.go
Comment (vgonkivs) [reply blob/service.go:180]: `IsBlobNamespace` was added by me, so I can add any message you want 🙂

File: header/pb/extended_header.proto
Comment (vgonkivs) [reply header/pb/extended_header.proto:13]: no, why? we are using an interface type there.

File: share/eds/byzantine/byzantine.go
Comment (vgonkivs) [reply share/eds/byzantine/byzantine.go:48]: In case there is something wrong with the share we can try next one as long as we need only half of them.

File: blob/parser.go
Comment (Wondertan) [reply blob/parser.go:83]: Hmm, we still planned to return ErrBlobNotFound in case of singular `GetBlob`, while we removed it for multiple.

File: blob/blobtest/testing.go
Comment (Wondertan) [inline blob/blobtest/testing.go:15]: Move to https://github.com/celestiaorg/go-square/pull/109?

File: blob/blobtest/testing.go
Comment (Wondertan) [inline blob/blobtest/testing.go:27]: This is gonna make the test flaky; if the rand bytes turn out to be reserved, it will error. Loop until it's valid instead?

File: blob/blob.go
Comment (Wondertan) [inline blob/blob.go:18]: `appconst.LatestVersion` so we don't have to bump it all the time

**Review by Wondertan — COMMENTED** [review]
Massive

File: blob/blob.go
Comment (vgonkivs) [reply blob/blob.go]: made it uint8

File: api/docgen/examples.go
Comment (vgonkivs) [reply api/docgen/examples.go]: named `libshare`

File: blob/helper.go
Comment (vgonkivs) [reply blob/helper.go:12]: to ensure they have correct namespace order

File: blob/helper.go
Comment (vgonkivs) [reply blob/helper.go]: do you mean to make in go-square?

File: share/eds/proofs_cache.go
Comment (vgonkivs) [reply share/eds/proofs_cache.go:321]: previously, share was an alias to []byte so it worked fine. Since now it is `struct{[]byte}`, I decided to not break this function. I will come back to this function after all fixes to ensure that new share type will not significantly break the call stack

File: share/eds/read.go
Comment (vgonkivs) [reply share/eds/read.go]: True.

File: share/shwap/share.go
Comment (vgonkivs) [reply share/shwap/share.go:31]: cc @cmwaters

File: blob/helper.go
Comment (Wondertan) [reply blob/helper.go]: Yes, starting from an issue proposing this method

File: share/eds/read.go
Comment (Wondertan) [reply share/eds/read.go]: Care for an issue in go-squre proposing Share constructor to return value not a type?

File: blob/parser.go
Comment (Wondertan) [reply blob/parser.go]: @vgonkivs

File: blob/helper.go
Comment (vgonkivs) [reply blob/helper.go]: already made a method  in go-square 🙂

File: nodebuilder/da/service.go
Comment (vgonkivs) [reply nodebuilder/da/service.go:220]: Agreed with @tzdybal to leave it `v0`

File: store/file/ods.go
Comment (vgonkivs) [reply store/file/ods.go]: yes https://github.com/celestiaorg/go-square/blob/8e9b275b5c6db9d82180ef2af01171fe12859d74/share/namespace.go#L24

Comment (vgonkivs) [reply store/file/ods.go]: yeap:
1) first 29 bytes contains share namespace: https://github.com/celestiaorg/go-square/blob/8e9b275b5c6db9d82180ef2af01171fe12859d74/share/share.go#L38
2) first byte contains ns version:
https://github.com/celestiaorg/go-square/blob/8e9b275b5c6db9d82180ef2af01171fe12859d74/share/namespace.go#L24

File: blob/parser.go
Comment (Wondertan) [reply blob/parser.go]: Discussed in DM. This does not remove ErrBlobNotFound usage entirely. Fun fact that this case shouldn't be possible in practice as it would be mean either validators made malformed block or GetSharesByNamespace proofs were wrong.

File: store/file/ods.go
Comment (Wondertan) [reply store/file/ods.go]: I recommend using `share.NewNamespaceFromBytes` as it doesn't allocate bytes slice inside

Comment (Wondertan) [reply store/file/ods.go]: Please make sure we do that everywhere we construct a namespace from bytes.

File: share/eds/read.go
Comment (Wondertan) [reply share/eds/read.go]: https://github.com/celestiaorg/go-square/issues/116

File: go.mod
Comment (evan-forbes) [inline go.mod:10]: we have v3.0.0-rc0 now

tried to open a PR, but didn't want to open on @vgonkivs 's fork and github doesn't keep track of that fork for some reason

Comment (vgonkivs) [reply go.mod:10]: I will bump it here. Thanks @evan-forbes

File: store/file/ods.go
Comment (vgonkivs) [reply store/file/ods.go]: done

**Review by Wondertan — DISMISSED** [review]
Massive. Good job

File: share/shwap/row_namespace_data_test.go
Comment (cristaloleg) [inline share/shwap/row_namespace_data_test.go:44]: Should this test be uncommented?

**Review by cristaloleg — APPROVED** [review]
LGTM

File: share/shwap/row_namespace_data_test.go
Comment (vgonkivs) [reply share/shwap/row_namespace_data_test.go:44]: nope.  Absence proof should be supported in shwap

**Review by Wondertan — APPROVED** [review]

**Comment by evan-forbes** [comment]
> We are not backward compatible since the app version has changed

quick question, does this mean celestia node nodes can't bump until the app version is changed?

**Comment by vgonkivs** [comment]
> quick question, does this mean celestia node nodes can't bump until the app version is changed?

I meant that that nodes can't sync if they won't upgrade to v3 once we launch v3 compatible network(bc header version will be changed to v3 and header validation will fail)

**Comment by vgonkivs** [comment]
Waiting until all shwap related changes will be merged to main to resolve all merge conflicts

**Comment by Wondertan** [comment]
@vgonkivs, there is a new conflict to be resolved

**Review by Wondertan — APPROVED** [review]

**Review by renaynay — APPROVED** [review]

**Review by walldiss — APPROVED** [review]

**Pull request merged**
